File: solutions/python/tries/map-sum-pairs.py
# LeetCode 677 — Map Sum Pairs (Medium)
# Category: Tries · Approach: Trie of running sums
# Time: O(k) per sum | Space: O(total keys)
# Source: https://leetcode.com/problems/map-sum-pairs/

class MapSum:
    def __init__(self) -> None:
        self.values: dict[str, int] = {}
        self.prefixes: dict[str, int] = {}

    def insert(self, key: str, val: int) -> None:
        delta = val - self.values.get(key, 0)
        self.values[key] = val
        prefix = ""
        for ch in key:
            prefix += ch
            self.prefixes[prefix] = self.prefixes.get(prefix, 0) + delta

    def sum(self, prefix: str) -> int:
        return self.prefixes.get(prefix, 0)

File: solutions/typescript/tries/map-sum-pairs.ts
// LeetCode 677 — Map Sum Pairs (Medium)
// Category: Tries · Approach: Map + prefix scan
// Time: O(k) per sum | Space: O(total keys)
// Source: https://leetcode.com/problems/map-sum-pairs/

class MapSum {
  private values = new Map<string, number>();

  insert(key: string, val: number): void {
    this.values.set(key, val);
  }

  sum(prefix: string): number {
    let total = 0;
    for (const [key, val] of this.values) {
      if (key.startsWith(prefix)) total += val;
    }
    return total;
  }
}
